plugin-base/src/main/java/io/everytrade/server/plugin/impl/everytrade/BitmexConnector.java

```diff
@@ -0,0 +1,208 @@
+            actualLastTransactionId = String.format("%s%s%s", lastUserTradeOffset, TX_SPLITER, lastUserTradeId);
+        } else {
+            actualLastTransactionId = lastTransactionId;
+        }
+
+        final ParseResult parseResult = XChangeConnectorParser.getParseResult(userTrades, SupportedExchange.BITMEX);
+
+        return new DownloadResult(parseResult, actualLastTransactionId);
+    }
+
+    private List<UserTrade> download(TransactionIdentifier lastTransactionId, TradeService tradeService) {
+        final BitmexTradeHistoryParams tradeHistoryParams
+            = (BitmexTradeHistoryParams) tradeService.createTradeHistoryParams();
+        tradeHistoryParams.setLimit(MAX_TXS_PER_REQUEST);
+        final List<UserTrade> userTrades = new ArrayList<>();
+        TransactionIdentifier lastDownloadedTx
+            = new TransactionIdentifier(lastTransactionId.offset, lastTransactionId.id);
+        int sentRequests = 0;
+        final long requestStart = System.currentTimeMillis();
+
+        while (sentRequests < MAX_REQUESTS_COUNT) {
+            if (sentRequests >= MAX_REQUESTS_COUNT_PER_MINUTE) {
+                sleep(sentRequests, requestStart);
+            }
+            tradeHistoryParams.setOffset(lastDownloadedTx.offset);
+            final List<UserTrade> userTradesBlock;
+            try {
+                userTradesBlock = tradeService.getTradeHistory(tradeHistoryParams).getUserTrades();
+            } catch (IOException e) {
+                throw new IllegalStateException("User trade history download failed. ", e);
+            }
+            final List<UserTrade> userTradesToAdd;
+            final int duplicateTxIndex = findDuplicate(lastDownloadedTx.id, userTradesBlock);
+            if (duplicateTxIndex > -1) {
+                if (duplicateTxIndex < userTradesBlock.size() - 1) {
+                    userTradesToAdd = userTradesBlock.subList(duplicateTxIndex + 1, userTradesBlock.size());
+                } else {
+                    userTradesToAdd = List.of();
+                }
+            } else {
+                userTradesToAdd = userTradesBlock;
+            }
+
+            if (userTradesToAdd.isEmpty()) {
+                break;
+            }
+
+            final UserTrade userTradeLast = userTradesToAdd.get(userTradesToAdd.size() - 1);
+            final long actualOffset = lastTransactionId.offset + userTradesBlock.size();
+            lastDownloadedTx = new TransactionIdentifier(actualOffset, userTradeLast.getId());
+
+            userTrades.addAll(userTradesToAdd);
+            ++sentRequests;
+        }
+
+        return userTrades;
+    }
+
+    private void sleep(int sentRequests, long requestStart) {
+        long sleepTime;
+        if (sentRequests == MAX_REQUESTS_COUNT_PER_MINUTE) {
+            final long maxRequestsDuration = System.currentTimeMillis() - requestStart;
+            sleepTime = 60_000L - maxRequestsDuration;
```

Comment:
there should be an upper bound for waiting. Let's say we blow our request budget in 1 second. Sleeping for 59 seconds is unacceptable.

Reply:
Let's start with max 5 seconds wait.